develop/2017-05-29-fgrehm-analysis-on-meals-based-on-presence-information.py

```diff
@@ -0,0 +1,467 @@
+    how='left',
+    on='document_id'
+)
+print("Meals with timestamps:", len(meals[~meals.timestamps.isnull()]))
+
+
+# In[11]:
+
+meals = meals[[
+    'document_id',
+    'applicant_id',
+    'congressperson_document',
+    'congressperson_id',
+    'congressperson_name',
+    'congressperson_state',
+    'party',
+    'year',
+    'issue_date',
+    'total_net_value',
+    'company_name',
+    'company_state',
+    'timestamps'
+]]
+
+
+# ## Meals outside DF while the congressperson was present in a session
+# 
+# There are at least two ways we can identify that:
+# * Making an approximation of the total time the congressperson was in Brasilia (like more than X hours) and checking for reimbursements on days they were there for a long period of time
+# * Leveraging OCR data from receipts and grabbing their timestamps
+
+# In[12]:
+
+deputies = pd.read_csv('../data/2017-04-19-deputies.xz', low_memory=False)
+print("Total deputies:", len(deputies))
+
+sessions = pd.read_csv('../data/2017-04-19-session-start-times.xz', low_memory=False)
+sessions['date'] = pd.to_datetime(sessions['date'], format="%Y-%m-%dT%H:%M:%S").dt.date
+sessions['started_at'] = pd.to_datetime(sessions['started_at'], format="%Y-%m-%dT%H:%M:%S")
+print("Session records:", len(sessions))
+
+presences = pd.read_csv('../data/2017-04-19-presences.xz', low_memory=False)
+presences['date'] = pd.to_datetime(presences['date'], format="%Y-%m-%dT%H:%M:%S").dt.date
+presences.sort_values('date', ascending=False)
+print("Presence records:", len(presences))
+
+presences = presences.query('presence == "Present"')
+print("Presence records with presence confirmed:", len(presences))
+
+# Match presence with session and deputy info
+presences = pd.merge(presences, sessions, on=['date', 'session'])
+print("Presence records with session matched:", len(presences))
+
+presences = pd.merge(deputies, presences, on='congressperson_document')
+print("Presence records with deputies matched:", len(presences))
+
+
+# In[13]:
+
+presences['first_session_at'] = presences['started_at']
+presences['last_session_at'] = presences['started_at']
+presences['total_sessions'] = 0
+aggregations = {
+    'first_session_at': 'min',
+    'last_session_at': 'max',
+    'total_sessions': 'count',
+    'session': lambda x: "{%s}" % ', '.join(x.astype(str)),
+}
+presences = presences.groupby(['congressperson_id', 'date'], as_index=False).agg(aggregations)
+print("Confirmed presences grouped by date and congressperson:", len(presences))
+
+
+# In[17]:
+
+meals_outside_df = meals.query('company_state != "DF"')
+meals_outside_df_while_in_df = pd.merge(
+    left=presences, 
+    right=meals_outside_df, 
+    left_on=['congressperson_id', 'date'],
+    right_on=['congressperson_id', 'issue_date']
+)
+print("Meals outside DF on days that the congressperson was there:", len(meals_outside_df_while_in_df))
+
+meals_outside_df_while_in_df = meals_outside_df_while_in_df.query('company_name != "GOL LINHAS AEREAS S.A."')
```

Comment:
> If so, isn't there a chance to exist other airlines in the dataset?

Probably yes, I just got gol out of the way since in previous analysis they were generating some noise, if they come up at the top of the suspicious reimbursements I'm going to try to clean those up as well.